tensorflow_probability/python/experimental/mcmc/sample_fold.py

```diff
@@ -133,3 +139,174 @@ def sample_fold(
     return (reduction_results,
             end_state,
             final_kernel_results.inner_results.inner_results)
+
+
+class StatesAndTrace(
```

Comment:
Why copy these from tfp.mcmc instead of just reusing them?

Reply:
good question, reusing is probably better